com.ca.lsp.cobol/lsp-service-cobol/src/main/java/com/ca/lsp/cobol/service/utils/WorkspaceFileService.java

```diff
@@ -0,0 +1,254 @@
+/*
+ * Copyright (c) 2020 Broadcom.
+ * The term "Broadcom" refers to Broadcom Inc. and/or its subsidiaries.
+ *
+ * This program and the accompanying materials are made
+ * available under the terms of the Eclipse Public License 2.0
+ * which is available at https://www.eclipse.org/legal/epl-2.0/
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ *
+ * Contributors:
+ *    Broadcom, Inc. - initial API and implementation
+ */
+package com.ca.lsp.cobol.service.utils;
+
+import com.google.inject.Singleton;
+import lombok.extern.slf4j.Slf4j;
+import org.apache.commons.io.FilenameUtils;
+
+import javax.annotation.Nonnull;
+import javax.annotation.Nullable;
+import java.io.File;
+import java.io.IOException;
+import java.io.UnsupportedEncodingException;
+import java.net.URI;
+import java.net.URISyntaxException;
+import java.net.URLDecoder;
+import java.nio.file.*;
+import java.util.List;
+import java.util.stream.Stream;
+
+import static java.lang.String.format;
+import static java.nio.file.Files.readAllLines;
+import static java.nio.file.StandardOpenOption.WRITE;
+import static java.util.Arrays.asList;
+import static java.util.Collections.emptyList;
+import static java.util.stream.Collectors.toList;
+
+/**
+ * This service implements API for low-level file systems access. It mainly oriented to work with
+ * copybooks, especially with .copybooks folder.
+ */
+@Singleton
+@Slf4j
+public class WorkspaceFileService implements FileSystemService {
+  private static final List<String> ALLOWED_EXTENSIONS = asList("cpy", "cbl", "cobol", "cob");
+  private static final String COPYBOOKS_FOLDER = ".copybooks";
+  private static final String C4Z_FOLDER = ".c4z";
+
+  @Override
+  public boolean fileExists(@Nullable Path file) {
+    return file != null && file.toFile().exists();
+  }
+
+  @Override
+  public void writeInFile(
+      @Nonnull Path targetPath,
+      @Nonnull String content,
+      @Nullable StandardOpenOption standardOpenOption) {
+    try {
+      Files.write(targetPath, (content + "\n").getBytes(), WRITE, standardOpenOption);
+    } catch (IOException e) {
+      log.error("Error writing to file: " + targetPath, e);
+    }
+  }
+
+  @Nonnull
+  @Override
+  public String decodeURI(@Nonnull String uri) {
+    try {
+      return URLDecoder.decode(uri, "UTF-8");
+    } catch (UnsupportedEncodingException e) {
+      log.error("Cannot decode URI: " + uri, e);
+      return uri;
+    }
+  }
+
+  @Nonnull
+  @Override
+  public List<String> interpretPaths(@Nonnull List<Object> settings) {
+    return settings.stream()
+        .map(Object::toString)
+        .map(
+            path ->
+                path.substring(path.indexOf(COPYBOOKS_FOLDER) + COPYBOOKS_FOLDER.length())
+                    .replace("/", filesystemSeparator())
```

Comment:
This line doesn't cover all cases. It converts Unix separators to Windows, but leave Windows separators unchanged for Unix.